### PR TITLE
Relax int type restriction

### PR DESCRIPTION
**Pull request by okuta**
Current int type restriction is too strict.
This PR is relax it.

### CODE REVIEW

File: chainer/functions/connection/embed_id.py [hunk 1]
```diff
@@ -81,16 +81,17 @@ def forward(self, inputs):
...
                     'embed_id_bwd')(
                         gy, xp.expand_dims(x, -1), gW.shape[1], gW)
             else:
                 cuda.elementwise(
-                    'T gy, int32 x, int32 n_out, int32 ignore', 'raw T gW',
+                    'T gy, S x, S n_out, S ignore', 'raw T gW',
                     '''
                     if (x != ignore) {
-                      int w_ind[] = {x, i % n_out};
+                      ptrdiff_t w_ind[] = {x, i % n_out};
```
Comment (niboshi) [inline chainer/functions/connection/embed_id.py]: Ditto. (`ptrdiff_t` -> `S`)

File: chainer/functions/evaluation/binary_accuracy.py [hunk 2]
```diff
@@ -45,7 +43,7 @@ def binary_accuracy(y, t):
 
         t (:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
         :class:`cupy.ndarray`):
-            Array holding an int32 vector of ground truth labels.
+            Array holding an signed integer vector of ground truth labels.
```
Comment (niboshi) [inline chainer/functions/evaluation/binary_accuracy.py]: an -> a

File: chainer/functions/loss/sigmoid_cross_entropy.py [hunk 3]
```diff
@@ -75,8 +75,8 @@ def sigmoid_cross_entropy(
         x (Variable): A variable object holding a matrix whose (i, j)-th
             element indicates the unnormalized log probability of the j-th unit
             at the i-th example.
-        t (Variable): Variable holding an int32 vector of ground truth labels.
-            If ``t[i] == -1``, corresponding ``x[i]`` is ignored.
+        t (Variable): Variable holding an signed integer vector of ground truth
```
Comment (niboshi) [inline chainer/functions/loss/sigmoid_cross_entropy.py]: ditto. (a -> an)

File: chainer/functions/loss/softmax_cross_entropy.py [hunk 4]
```diff
@@ -294,7 +294,7 @@ def softmax_cross_entropy(
             dimensions is greater than 2.
         t (:class:`~chainer.Variable` or :class:`numpy.ndarray` or \
         :class:`cupy.ndarray`):
-            Variable holding an :class:`numpy.int32` vector of ground truth
+            Variable holding an signed integer vector of ground truth
```
Comment (niboshi) [inline chainer/functions/loss/softmax_cross_entropy.py]: ditto. (an -> a)

File: tests/chainer_tests/functions_tests/array_tests/test_permutate.py [hunk 5]
```diff
@@ -10,19 +10,20 @@
...
-    {'shape': (3,), 'dtype': 'f', 'axis': -1, 'inv': True},
-    {'shape': (3, 4), 'dtype': 'd', 'axis': 1, 'inv': True},
-    {'shape': (3, 4, 5), 'dtype': 'f', 'axis': 2, 'inv': False},
-)
+@testing.parameterize(*testing.product_dict(
+    [{'shape': (3,), 'dtype': 'f', 'axis': 0, 'inv': False},
+     {'shape': (3,), 'dtype': 'f', 'axis': -1, 'inv': True},
+     {'shape': (3, 4), 'dtype': 'd', 'axis': 1, 'inv': True},
+     {'shape': (3, 4, 5), 'dtype': 'f', 'axis': 2, 'inv': False}],
+    [{'label_dtype': numpy.int}, {'label_dtype': numpy.int32}]
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/array_tests/test_permutate.py]: How about other int types (int8, int16, int64)?

File: tests/chainer_tests/functions_tests/connection_tests/test_embed_id.py [hunk 6]
```diff
@@ -11,16 +11,18 @@
...
-    {'x_data': [0, 1, -1], 'ignore_label': -1},
-    {'x_data': [[0, 1, -1], [-1, 0, 1]], 'ignore_label': -1},
-)
+@testing.parameterize(*testing.product_dict(
+    [{'x_data': [0, 1, 0], 'ignore_label': None},
+     {'x_data': [[0, 1, 0], [1, 0, 1]], 'ignore_label': None},
+     {'x_data': [0, 1, -1], 'ignore_label': -1},
+     {'x_data': [[0, 1, -1], [-1, 0, 1]], 'ignore_label': -1}],
+    [{'label_dtype': numpy.int},
+     {'label_dtype': numpy.int32}]
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/connection_tests/test_embed_id.py]: Ditto. (other int types)

File: tests/chainer_tests/functions_tests/evaluation_tests/test_accuracy.py [hunk 7]
```diff
@@ -50,7 +49,9 @@ def accuracy(x, t, ignore_label):
          {'ignore_label': 0, 't_data': 'zero'}],
         [{'dtype': numpy.float16},
          {'dtype': numpy.float32},
-         {'dtype': numpy.float64}]
+         {'dtype': numpy.float64}],
+        [{'label_dtype': numpy.int},
+         {'label_dtype': numpy.int32}]
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/evaluation_tests/test_accuracy.py]: Ditto. (other int types)

File: tests/chainer_tests/functions_tests/evaluation_tests/test_binary_accuracy.py [hunk 8]
```diff
@@ -7,19 +7,20 @@
...
 from chainer import testing
 from chainer.testing import attr
-from chainer.testing import condition
 from chainer.utils import type_check
 
 
 @testing.parameterize(*testing.product({
     'shape': [(9, 11), (99,)],
     'dtype': [numpy.float16, numpy.float32, numpy.float64],
+    'label_dtype': [numpy.int, numpy.int32],
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/evaluation_tests/test_binary_accuracy.py]: Ditto. (other int types)

File: tests/chainer_tests/functions_tests/evaluation_tests/test_classification_summary.py [hunk 9]
```diff
@@ -64,33 +63,44 @@ def make_ground_truth(label_num, shape, ignore_label, has_ignore_label):
...
         [{'beta': 1.0},
          {'beta': 2.0}],
         [{'label_num': 3},
          {'label_num': None}],
         [{'ignore_label': -1},
          {'ignore_label': 0}],
         [{'has_ignore_label': True},
-         {'has_ignore_label': False}]
+         {'has_ignore_label': False}],
+        [{'label_dtype': numpy.int32}]
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/evaluation_tests/test_classification_summary.py]: Ditto. (other int types)

File: tests/chainer_tests/functions_tests/evaluation_tests/test_classification_summary.py [hunk 10]
```diff
@@ -64,33 +63,44 @@ def make_ground_truth(label_num, shape, ignore_label, has_ignore_label):
...
+        [{'y_shape': (100, 3), 't_shape': (100,)}],
+        [{'dtype': numpy.float16},
+         {'dtype': numpy.float32},
+         {'dtype': numpy.float64}],
+        [{'beta': 1.0}],
+        [{'label_num': 3}],
+        [{'ignore_label': 0}],
+        [{'has_ignore_label': True}],
+        [{'label_dtype': numpy.int},
+         {'label_dtype': numpy.int32}]
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/evaluation_tests/test_classification_summary.py]: Ditto. (other int types)

File: tests/chainer_tests/functions_tests/loss_tests/test_contrastive.py [hunk 11]
```diff
@@ -16,7 +16,7 @@
 @testing.parameterize(
     *testing.product({
         'batchsize': [5, 10], 'input_dim': [2, 3], 'margin': [1, 2],
-        'reduce': ['mean', 'no']
+        'reduce': ['mean', 'no'], 'label_dtype': [numpy.int32, numpy.int32]
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/loss_tests/test_contrastive.py]: Ditto. (other int types)

File: tests/chainer_tests/functions_tests/loss_tests/test_hinge.py [hunk 12]
```diff
@@ -12,19 +12,18 @@
...
-@testing.parameterize(
-    *testing.product(
-        {'reduce': ['no', 'mean'],
-         'norm': ['L1', 'L2']}
-    )
-)
+@testing.parameterize(*testing.product({
+    'reduce': ['no', 'mean'],
+    'norm': ['L1', 'L2'],
+    'label_dtype': [numpy.int, numpy.int32],
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/loss_tests/test_hinge.py]: Ditto. (other int types)

File: tests/chainer_tests/functions_tests/loss_tests/test_sigmoid_cross_entropy.py [hunk 13]
```diff
@@ -15,21 +15,23 @@
...
-    {'shape': (8, 7), 'normalize': False},
-    {'shape': (8, 7), 'normalize': True, 'ignore_all': True},
+    {'shape': (8, 7), 'normalize': True, 'label_dtype': numpy.int32},
+    {'shape': (8, 7), 'normalize': False, 'label_dtype': numpy.int32},
+    {'shape': (8, 7), 'normalize': True, 'ignore_all': True,
+     'label_dtype': numpy.int32},
     # too large shape causes int32 -> float64 issue
-    {'shape': (65536, 1), 'normalize': False},
+    {'shape': (65536, 1), 'normalize': False, 'label_dtype': numpy.int32},
+    {'shape': (8, 7), 'normalize': True, 'label_dtype': numpy.int}
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/loss_tests/test_sigmoid_cross_entropy.py]: How about adding some cases to test other int types?

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 14]
```diff
@@ -21,6 +21,7 @@
     'dtype': [numpy.float32],
     'weight_apply': [False, True],
     'enable_double_backprop': [False, True],
+    'label_dtype': [numpy.int32],
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Ditto. (other int types)

File: tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py [hunk 15]
```diff
@@ -29,6 +30,7 @@
     'dtype': [numpy.float16, numpy.float32, numpy.float64],
     'weight_apply': [False, True],
     'enable_double_backprop': [False, True],
+    'label_dtype': [numpy.int, numpy.int32],
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/loss_tests/test_softmax_cross_entropy.py]: Ditto. (other int types)

File: chainer/functions/connection/embed_id.py [hunk 16]
```diff
@@ -81,16 +81,17 @@ def forward(self, inputs):
         else:
             if self.ignore_label is None:
                 cuda.elementwise(
-                    'T gy, int32 x, int32 n_out', 'raw T gW',
-                    'int w_ind[] = {x, i % n_out}; atomicAdd(&gW[w_ind], gy)',
+                    'T gy, S x, S n_out', 'raw T gW',
+                    'ptrdiff_t w_ind[] = {x, i % n_out};'
```
Comment (niboshi) [inline chainer/functions/connection/embed_id.py]: How about `ptrdiff_t` -> `S`? It's guaranteed to be representable by `S`.

Comment (okuta) [reply chainer/functions/connection/embed_id.py]: `ptrdiff_t` is better for CuPy.

File: chainer/functions/connection/embed_id.py (hunk 1 above)
Comment (okuta) [reply chainer/functions/connection/embed_id.py]: ditto.

File: chainer/functions/connection/embed_id.py (hunk 16 above)
Comment (niboshi) [reply chainer/functions/connection/embed_id.py]: OK, I agree with you because it's used as an index.

**Comment by niboshi** [comment]
Please fix AppVeyor result.

**Comment by okuta** [comment]
I fixed.

**Comment by okuta** [comment]
I resolved conflict.

File: chainer/functions/evaluation/classification_summary.py [hunk 17]
```diff
@@ -43,6 +42,7 @@ def check_type_forward(self, in_types):
     def forward(self, inputs):
         xp = cuda.get_array_module(*inputs)
         y, t = inputs
+        t = t.astype('i', copy=False)
```
Comment (niboshi) [inline chainer/functions/evaluation/classification_summary.py]: Is this conversion necessary?
I think the actual type for `'i'` is environment-dependent.

Comment (okuta) [reply chainer/functions/evaluation/classification_summary.py]: Yes. Some Numpy functions require int32 on Windows.

Comment (niboshi) [reply chainer/functions/evaluation/classification_summary.py]: I see. Can you write a brief comment on this line?

Comment (okuta) [reply chainer/functions/evaluation/classification_summary.py]: OK

**Comment by niboshi** [comment]
Please fix conflict.

**Comment by okuta** [comment]
I resolved conflict.

**Comment by niboshi** [comment]
(Jenkins has passed. Waiting for travis)

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**